add qe workflows and outline evo updates 🪂

SWeav02 opened the pull request:
This PR is related to part 1 of discussion #695. The goal is to create a new default workflow for the evolutionary search app that uses quantum espresso instead of vasp. The steps as I see them are:

- [x] Add `_get_clean_structure` method used in `VaspWorkflow` to `PwscfWorkflow`.
- [x] Add smearing keyword modifier for quantum espresso equivalent to `keyword_modifier_smart_ismear`
- [x] Create relaxation workflows of varying quality (00-04)
- [x] Create static workflow of higher quality
- [x] Create nested workflow
- [x] Test and bugfix

There may be other things worth doing like consolidating copied methods used in both the vasp and quantum espresso apps. It may also be good to add error handlers for QE.

Since the initial workflows are complete, I'll keep adding new items from #695 that I'm working on:
- [x] Allow custom workflows for staged calculation
Users can now use the "StagedWorkflow" mix-in which allows one to set a list of DFT workflows (either vasp or QE). 
So long as they use the StaticEnergy or Relaxation tables and the calculation outputs the values needed for the 
desired fitness function, this should work fine.

Moving to PRs after this one:
- [ ] Move similarity check to within postgres database
- [ ] Allow custom fitness function other than minimum E/atom
- [ ] Add steady state benchmarking to dynamically update steady state source ratios
- [ ] Add monitoring dashboard to ease visualization as search progresses

SWeav02 committed:
Create quantum espress quality00 relaxation

Created a quantum espresso relaxation workflow of quality00 comparable to the equivalent in the vasp app.

SWeav02 committed:
Added relaxation workflows of varying quality

Qualities 00-04 added. Also altered ecutrho and ecutwfc keyword modifiers to be dynamic to allow for automatic varying quality of encut.

jacksund commented:
> There may be other things worth doing like consolidating copied methods used in both the vasp and quantum espresso apps

I could see a `StructureWorkflow` mixin being helpful for things like the `_get_clean_structure` method -- and having VASP + QE base workflows pulling inherit this mixin. Also a `StagedWorkflow` class could be helpful to remove some boilerplate code. I'll leave these up to you though. Maybe a 2nd refactor PR can follow this one

>  It may also be good to add error handlers for QE.

I'm not sure other packages have error handlers, but you could maybe check out custodian/pymatgen/atomate and ase packages to see if they check for common errors and borrow those. Error handling would be lower priority for now because the target is evo searches

SWeav02 committed:
Important update to doc-strings :sunglasses:

SWeav02 commented:
> I could see a `StructureWorkflow` mixin being helpful for things like the `_get_clean_structure` method -- and having VASP + QE base workflows pulling inherit this mixin. Also a `StagedWorkflow` class could be helpful to remove some boilerplate code. I'll leave these up to you though. Maybe a 2nd refactor PR can follow this one

I don't think this PR will take much more time, so hopefully I can take a stab at this tomorrow. It would be convenient going forward.

SWeav02 committed:
Doc typos and command bug fixes

Several issues in the doc strings for relaxation calculations. Also fixed a bug where docker commands run in directories with a space would throw an error

SWeav02 committed:
Add abstract StagedWorkflow

Created a staged workflow class for running multiple dft calculations following eachother. Added staged relaxation workflow for QE

SWeav02 committed:
Add abstract StructureWorkflow

Created a structure workflow class that can be mixed into workflows that might want to clean a structure in their setup (e.g. VaspWorkflow, QuantumEspressoWorkflow)

SWeav02 commented:
@jacksund The new QuantumEspresso staged workflow is working, but without parallelization it takes a while even on primitive NaCl. I tried running with mpirun, but this throws an error:

```
mpirun has detected an attempt to run as root.
Running at root is *strongly* discouraged as any mistake (e.g., in
defining TMPDIR) or bug can result in catastrophic damage to the OS
file system, leaving your system in an unusable state.

You can override this protection by adding the --allow-run-as-root
option to your cmd line. However, we reiterate our strong advice
against doing so - please do so at your own risk.
```

From what I gather this is related to the docker image not having a user. Should we build the docker images with a user or is there a better way to implement parallelization?

I've also never really used QE, so the settings I chose are just an attempt to mirror the VASP settings in `relaxation.vasp.staged`. There might be some QE settings I don't know much about that greatly effect the speed (other than standard energy cutoffs, k spacing, convergence criteria). Any suggestions?

SWeav02 committed:
Added (fully untested/finished) structure similarity

Added a column for fingerprints in the Structure model that is calculated when a structure is added. Also started adding a method to compare using pgvector

SWeav02 committed:
Rework _default_database

Calling simmate database reset always created a .sqlite3 local database. I've reworked it to check for a database.yaml file and use whatever backend is defined there. This is not working yet, as I keep getting an error saying "samtest" (the name of my testing db) does not exist from django's attempts to connect.

jacksund commented on src/simmate/configuration/load_settings.py:
```diff
@@ -283,15 +283,43 @@ def default_settings(self) -> dict:
-            "engine": "django.db.backends.sqlite3",
-            "name": db_filename,
-        }
+        # !!! There may be a better way to do this, but this works for me
+        # We want to dynamically set the database backend in case the user is using
+        # a cloud database such as postgres
+        if Path(self.config_directory / f"{self.conda_env}-database.yaml").exists():
+            # if the user is in the (base) env or not using conda, then we will have a
+            # value of "-database.sqlite3", which is why we need strip() here.
+            db_filename = (
+                self.config_directory / f"{self.conda_env}-database.yaml".strip("-")
+            )
+            with open(self.config_directory / f"{self.conda_env}-database.yaml") as file:
+                yaml_file = yaml.safe_load(file)
+            default_db = {
+                "engine": yaml_file["default"]["ENGINE"],
+                "name": yaml_file["default"]["NAME"],
+                "user": yaml_file["default"]["USER"],
+                "password": yaml_file["default"]["PASSWORD"],
+                "port": yaml_file["default"]["PORT"],
+                "host": yaml_file["default"]["HOST"],
+            }
+            # try and add sslmode if it is there
+            try:
+                default_db["options"] = {"sslmode" : yaml_file["default"]["OPTIONS"]["sslmode"]}
+            except:
+                default_db["options"] = {"sslmode":"prefer"}
+            return default_db
+        
+        else:
+            # we want to use sqlite3 as our backend
+            # if the user is in the (base) env or not using conda, then we will have a
+            # value of "-database.sqlite3", which is why we need strip() here.
+            db_filename = (
+                self.config_directory / f"{self.conda_env}-database.sqlite3".strip("-")
+            )
+            return {
+                "engine": "django.db.backends.sqlite3",
+                "name": db_filename,
+            }
```
You'll have to undo this section. There is no longer a `*-database.yaml` file, but only a `settings.yaml` file that has everything! This is updated in the docs and I also gave more details to hyesol here:
https://github.com/jacksund/simmate/discussions/683#discussioncomment-10685167

SWeav02 replied:
This explains a lot :face_with_diagonal_mouth:
I was using the old `-database.yaml` file and it was of course just defaulting to sqlite3 instead. This was the first place in the code I saw that I could "fix" this so I went ahead and edited it. I've reverted this part back in the latest commit.

Unfortunately I still run into errors when running `simmate database reset`. One had to do with the psycopg2 connection to the maintenance database returning a None object, which I've fixed. Now I'm getting the following error from the `update_database` method  (full error in attachment):
```python
OperationalError: connection to server at "db-postgresql-nyc1-simmate-do-user-9931465-0.b.db.ondigitalocean.com" (206.189.189.161), port 25061 failed: FATAL:  no such database: samtest
```
When I investigate with DBeaver, the "samtest" database is shown in the database list, but it throws the same error if I try to connect to it directly. It seems like the database is successfully deleted, but something goes wrong with the creation. I've been searching for an explanation for a while but haven't found one. @jacksund any idea why this might be happening?

[error.txt](https://github.com/user-attachments/files/17466598/error.txt)

jacksund replied:
> When I investigate with DBeaver, the "samtest" database is shown in the database list, but it throws the same error if I try to connect to it directly.

> something goes wrong with the creation

Sounds like something is going on with the database and maybe permissions. Where is your database hosted and what version of postgres is it? Digital ocean -- if so, what are you seeing on the host platform? I just checked and confirmed resetting the db works with postgres when hosted with docker locally as well as AWS RDS.

As a fallback, you can create the database manually in the DO ui and then just use `simmate database update`, rather than `reset`

Also, you can include your new migrations in this MR (the ones that add the fingerprint column to a bunch of tables). Migrations are designed to be committed to the codebase (which is something I didn't know in earlier simmate versions)

SWeav02 replied:
>Sounds like something is going on with the database and maybe permissions. Where is your database hosted and what version of postgres is it? Digital ocean -- if so, what are you seeing on the host platform? I just checked and confirmed resetting the db works with postgres when hosted with docker locally as well as AWS RDS.
>
Thanks for checking! Permissions could make sense, although its odd it lets me delete full dbs if that's the case. I'm using our DigitalOcean server, but I don't have access to the ui. @scott-materials any chance you see anything on that end? If not, maybe we just try remaking the db in the DO ui and I'll run the `simmate database update` command instead.

>Also, you can include your new migrations in this MR (the ones that add the fingerprint column to a bunch of tables). Migrations are designed to be committed to the codebase (which is something I didn't know in earlier simmate versions)
>
Yeah sounds good. I was planning to once I got the postgres db going, but if I can't get it working soon I'll just go back to the sqlite3 db quick to generate them.

scott-materials replied:
> Thanks for checking! Permissions could make sense, although its odd it lets me delete full dbs if that's the case. I'm using our DigitalOcean server, but I don't have access to the ui. @scott-materials any chance you see anything on that end? If not, maybe we just try remaking the db in the DO ui and I'll run the `simmate database update` command instead.

@SWeav02 Is this still an issue?

jacksund commented on src/simmate/database/base_data_types/structure.py:
```diff
@@ -271,3 +284,85 @@ def to_toolkit(self) -> ToolkitStructure:
         return ToolkitStructure.from_database_object(self)
+    
+    @classmethod
+    def filter_similarity(cls, structure, method="L2", cutoff=0.8):
```
This is a good start. I have a more developed version of this with several improvements that I'll share once you flag the MR for review. If this is working for you now, just leave it -- if it's giving you issues, let me know and just hold off and let me refactor/update it

SWeav02 replied:
Sounds good. I've only been able to test it for the sqlite3 backend and it works fine for that, but I'll the postgres side once I get it working.

jacksund replied:
I removed you're `postgress_connect()` and will show why in a follow up MR. Effectively, [this is the preferred way](https://docs.djangoproject.com/en/5.1/topics/db/sql/) to do raw SQL queries in django, but there's an even better approach using annotations that I'll draft when I have time tomorrow

jacksund commented:
@SWeav02 @scott-materials This is kind of a cop out, but I think we should only support serial versions for docker-based builds -- and that's all I really considered when I built the QE image. It would be awesome if you guys are able to figure out parallelization, but I wouldn't get hung up on it. 

We should aim for a default workflow that can run a 5 atom system on 1 core in under 15 min (I'm just guessing at reasonable numbers here). And for the evo search tutorial, we can even have users switch to a `relaxation.quantum-espresso.evo-tutorial` that is SUPER low quality and good for running on laptops (maybe make it LDA w. only 2-3 separate relaxations). With a workflow like that in hand, we can easily get users to run a search on their laptop with 3 or so simmate workers.

@scott-materials any chance you're willing to share the VASP docker build? We are able to publish the dockerfile for it -- its just the vasp build files that need to be manually supplied. So I could have a `simmate-vasp` command that asks for the build files and then generates the VASP image from them -- and I probably only want a serial version here too. Here's an example repo that I found way back: https://github.com/ajjackson/vasp-docker

SWeav02 committed:
Remove rework _default_database

SWeav02 committed:
Merge remote-tracking branch 'origin/evolution_qespresso' into evolution_qespresso

SWeav02 committed:
Remove rework _default_database

SWeav02 commented:
> @SWeav02 @scott-materials This is kind of a cop out, but I think we should only support serial versions for docker-based builds -- and that's all I really considered when I built the QE image. It would be awesome if you guys are able to figure out parallelization, but I wouldn't get hung up on it.
> 
This is fine with me. I'm not too worried about it if it's primarily going to be used for the tutorial anyways.

> We should aim for a default workflow that can run a 5 atom system on 1 core in under 15 min (I'm just guessing at reasonable numbers here). And for the evo search tutorial, we can even have users switch to a `relaxation.quantum-espresso.evo-tutorial` that is SUPER low quality and good for running on laptops (maybe make it LDA w. only 2-3 separate relaxations). With a workflow like that in hand, we can easily get users to run a search on their laptop with 3 or so simmate workers.
> 
@jacksund Should I rework the default QE nested workflow to be more in line with this? Or keep it and make this as a new workflow?

jacksund commented:
> Should I rework the default QE nested workflow to be more in line with this? Or keep it and make this as a new workflow?

I'd keep what you have and have a separate "tutorial" workflow. The tutorial workflow would be nice for rapid local testing too

SWeav02 committed:
LDA Tutorial Workflow and Fingerprint Migrations

SWeav02 committed:
Fixed Major Typo

SWeav02 committed:
Add Band Info Parsing for QE

SWeav02 committed:
Add XML Parsing for QE relaxations

SWeav02 commented:
@jacksund when you've run calculations with QE before, were the energies you got out reasonable? For whatever reason, I get very large values (e.g. -877 eV/atom for NaCl). Vasp, with similar settings, gives a much more reasonable -3.4 eV/atom. I know the energy will vary with the program, but I would be shocked if they vary this much. I would say maybe it's something with the PPs we're using, but they're the first ones QE suggest on their website.

I'll keep messing around, but just wanted to know if you'd run into this or had ideas.

jacksund commented:
It's been a while, so I can't remember if this was happening to me or not... Energy does vary by program and potentials used, but yeah this is pretty large. The first thing I would check is hull energy of different crystals -- as those should be more comparable. I'll do a little testing later tonight

Is it alright if I update your branch to be in sync with the lastest `main` branch for testing? You might have to refresh your dev conda env because of a couple new deps -- `rdkit` and `umap (or you can just install these as extras)

SWeav02 commented:
> It's been a while, so I can't remember if this was happening to me or not... Energy does vary by program and potentials used, but yeah this is pretty large. The first thing I would check is hull energy of different crystals -- as those should be more comparable. I'll do a little testing later tonight
> 
Yeah I ran a quick hull diagram test for NaCl and got a large positive Ef for NaCl. I repeated it with vasp just to be sure I did it correctly (since it's been a while) and got something close to what they have on materials project (~-2eV/atom).
> Is it alright if I update your branch to be in sync with the lastest `main` branch for testing? You might have to refresh your dev conda env because of a couple new deps -- `rdkit` and `umap` (or you can just install these as extras)
>
Yep fine by me!

jacksund committed:
Merge branch 'main' into evolution_qespresso

jacksund commented:
alright, so I'm calling it for the night but here's where I got so far. Long story short, I think the issue is somewhere in my conversion of Ry to eV.

I don't think the issue is with the QE docker image because I can reproduce total energy values from others using QE. For example, there's [this tutorial](https://pranabdas.github.io/espresso/hands-on/scf) (and [the necessary UPF file for it](https://pseudopotentials.quantum-espresso.org/upf_files/Si.pz-vbc.UPF)), where I get a matching total energy in Ry.

I also tested to see if I could reproduce formation energies as seen on the Materials Project -- as relative energies should be consistent even with different potentials + software. So I pull these three structures:
- [Na](https://next-gen.materialsproject.org/materials/mp-10172)
- [Cl2](https://next-gen.materialsproject.org/materials/mp-22848)
- [NaCl](https://next-gen.materialsproject.org/materials/mp-22862)

And ran the following:
``` python
from simmate.workflows.utilities import get_workflow

wf_name = "static-energy.quantum-espresso.quality00"
workflow = get_workflow(wf_name)

files = ["Na.cif", "Cl2.cif", "NaCl.cif"]
for file in files:
    workflow.run(structure=file).result()

workflow.database_table.update_all_stabilities(wf_name)
```

Which gives the following results:
![image](https://github.com/user-attachments/assets/a6c043ab-848a-4e5a-9707-3601d37d8514)

So I got a predicted formation energy of `-1953 eV/atom` for NaCl while MatProj gives `-2038 meV/atom`. Note the meV -- meaning I am off by ~1000x. This might be a red herring, but [others](https://www.reddit.com/r/comp_chem/comments/dyrdkq/total_energy_differs_significantly_in_my_dft/) have seen this 1000x difference as well. Or maybe it's just a coincidence that it was ~1000x -- but testing a couple other systems should make that clear.

[This](https://github.com/SWeav02/simmate/blob/db2d853f244920479437bc8ea9b155f22223d152/src/simmate/apps/quantum_espresso/outputs/pwscf_xml.py#L96) is where I make the conversion from Ry to eV, but maybe I'm missing something...? Maybe the total energy is given in the xml as mRy?

jacksund committed:
fix db reset util and start reviewing

SWeav02 commented:
> I don't think the issue is with the QE docker image because I can reproduce total energy values from others using QE. For example, there's [this tutorial](https://pranabdas.github.io/espresso/hands-on/scf) (and [the necessary UPF file for it](https://pseudopotentials.quantum-espresso.org/upf_files/Si.pz-vbc.UPF)), where I get a matching total energy in Ry.
> 
Great, I'm glad we can rule that out! Thanks for running that.
> So I got a predicted formation energy of `-1953 eV/atom` for NaCl while MatProj gives `-2038 meV/atom`. Note the meV -- meaning I am off by ~1000x. This might be a red herring, but [others](https://www.reddit.com/r/comp_chem/comments/dyrdkq/total_energy_differs_significantly_in_my_dft/) have seen this 1000x difference as well. Or maybe it's just a coincidence that it was ~1000x -- but testing a couple other systems should make that clear.
> 
I'll test a couple more this morning after group meeting.
> [This](https://github.com/SWeav02/simmate/blob/db2d853f244920479437bc8ea9b155f22223d152/src/simmate/apps/quantum_espresso/outputs/pwscf_xml.py#L96) is where I make the conversion from Ry to eV, but maybe I'm missing something...? Maybe the total energy is given in the xml as mRy?
> 
Yesterday I spent a lot of time adding methods to this XML parser which is when I started noticing the energies. I'm pretty confident the conversion should be correct if the energy really is in Ry as they're labeled in the .out file. But the energies in the output are on the order of ~100 Ry/atom which is too high. Hopefully, it really is something as simple as the values being in mRy. I'll update with what I find either way.

jacksund commented:
sounds good. Let me know how things go. I would also look into energies reported in the UPF files as well as the different energies reported in the final xml. I'm seeing large energies reported in the Na and Cl UPF files, which we might want to exclude.

I'm also about to make one more commit that makes progress on the `filter_similarity` method. Will need to wait until later to finalize that feature though

jacksund committed:
outline filter_sim method. run black and isort.

jacksund commented:
> Move similarity check to within postgres database

Let's actually exclude this from this PR. We can keep the outline/notes that have been added, but introducing a dependency to the database is tricky for my production set up -- so I'd like to take my time on it and make sure things are correct before pushing new migrations

SWeav02 committed:
Fix error when not enough bands used QE

SWeav02 committed:
Merge remote-tracking branch 'origin/evolution_qespresso' into evolution_qespresso

SWeav02 commented:
This PR is ready for now. The goal of adding a QE staged workflow that works with the Evo algorithm is mostly complete, although we still have not figured out the energy bug. I've tested that the fixed composition workflow "works" with the new QE tutorial staged workflow as well and it seems to run fine, though of course gives nonsense energy values.

Next PR we'll start working on the parts of #695 that we didn't get to here.

jacksund commented:
sounds good. I'll do some cleanup and merge when I have some free time. I'll leave the vasp workflow as the default until we figure out QE too

SWeav02 commented:
> sounds good. I'll do some cleanup and merge when I have some free time. I'll leave the vasp workflow as the default until we figure out QE too

Great, thanks! I'll start working on the next steps.

SWeav02 committed:
Quick updates to QE workflow units

Forgot to commit this before opening PR for review

jacksund committed:
run black and remove db connect

jacksund committed:
remove fp col and mgs

scott-materials commented:
> @scott-materials any chance you're willing to share the VASP docker build? We are able to publish the dockerfile for it -- its just the vasp build files that need to be manually supplied. So I could have a `simmate-vasp` command that asks for the build files and then generates the VASP image from them -- and I probably only want a serial version here too. Here's an example repo that I found way back: https://github.com/ajjackson/vasp-docker

Here's the dockerfile and makefile.include for the serial build of VASP.  The dockerfile presumes we have a file called vasp.tar.gz in the same working directory as the dockerfile and that, when unzipped, vasp.tar.gz gives a folder called vasp in which the standard build files and directories are found (makefile.include, arch, bin, build, and others).  This structure is a bit different than the standard download from the vasp website, but easy to either modify or update the dockerfile accordingly.  I do this because proficient users are going to manually modify the standard vasp download, such as by applying the standard patches or applying custom fixes, such as the bug fix for the ELF code.

**dockerfile**
```bash
# Use Ubuntu 22.04 LTS as the base image
FROM ubuntu:22.04

# Set a working directory
WORKDIR /app

# Install necessary packages
RUN apt-get update && apt-get install -y \
    wget \
    build-essential \
    gfortran \
    libopenblas-dev \
    liblapacke-dev \
    rsync

#Install FFTW
RUN mkdir /app/fftw
RUN mkdir /app/fftw/install
WORKDIR /app/fftw
RUN wget https://www.fftw.org/fftw-3.3.10.tar.gz
RUN tar xvzf fftw-3.3.10.tar.gz
RUN mkdir /app/fftw/fftw-3.3.10/build
WORKDIR /app/fftw/fftw-3.3.10/build
RUN ../configure --prefix=/app/fftw/install
RUN make -j 20
RUN make install
ENV PATH /app/fftw/install/bin:$PATH
ENV INCLUDE /app/fftw/install/include:$INCLUDE
ENV LD_LIBRARY_PATH /app/fftw/install/lib:$LD_LIBRARY_PATH

# Install VASP
WORKDIR /app
COPY vasp.tar.gz /app/
RUN tar -xzvf vasp.tar.gz
WORKDIR /app/vasp
RUN make std
RUN cp bin/vasp_std bin/vasp
ENV PATH /app/vasp/bin:$PATH
RUN rm /app/vasp.tar.gz
```

**makefile.include:**
```bash
# Precompiler options
CPP_OPTIONS= -DHOST=\"LinuxGNU\" \
             -DMPI_NONE -DNGZhalf \
             -Duse_collective \
             -DCACHE_SIZE=4000 \
             -Davoidalloc \
             -Duse_bse_te \
             -Dtbdyn \
             -Duse_shmem

CPP_OPTIONS += -DLAPACK36
CPP_OPTIONS += -lpthread

CPP        = gcc -E -P -C -w $*$(FUFFIX) >$*$(SUFFIX) $(CPP_OPTIONS)

FC         = gfortran
FCL        = gfortran

FREE       = -ffree-form -ffree-line-length-none

FFLAGS     = -w -std=legacy
OFLAG      = -O2
OFLAG_IN   = $(OFLAG)
DEBUG      = -O0

BLAS       = /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.a
LAPACK     = /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.a
SCALAPACK  = 

LLIBS      = $(LAPACK) $(BLAS)

FFTW       ?= /app/fftw/install
LLIBS      += -L$(FFTW)/lib -lfftw3
INCS       = -I$(FFTW)/include

OBJECTS    = fftw3d.o  fft3dlib.o

OBJECTS_O1 += fftw3d.o
OBJECTS_O2 += fft3dlib.o

# For what used to be vasp.5.lib
CPP_LIB    = $(CPP)
FC_LIB     = $(FC)
CC_LIB     = gcc
CFLAGS_LIB = -O
FFLAGS_LIB = -O1
FREE_LIB   = $(FREE)

OBJECTS_LIB= linpack_double.o getshmem.o

# For the parser library
CXX_PARS   = g++

LIBS       += parser
LLIBS      += -Lparser -lparser -lstdc++

# Normally no need to change this
SRCDIR     = ../../src
BINDIR     = ../../bin
```

jacksund committed:
attempt matminer patch

jacksund committed:
update requests dep

jacksund committed:
remove fp from db loading

jacksund committed:
patch several unittests

jacksund committed:
Merge branch 'main' into evolution_qespresso
